File: Editorial/2464_Minimum_Subarrays_in_a_Valid_Split/JavaScript/1_Approach_Dynamic_Programming.js
const INF = 1e9;

function gcd(a, b) {
    if (b === 0) return a;
    return gcd(b, a % b);
}

var validSubarraySplit = function (nums) {
    const n = nums.length;
    const dp = new Array(n + 1).fill(INF);
    dp[0] = 0;
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= i; j++) {
            if (gcd(nums[i - 1], nums[j - 1]) > 1) {
                dp[i] = Math.min(dp[i], dp[j - 1] + 1);
            }
        }
    }
    return dp[n] === INF ? -1 : dp[n];
};

File: Editorial/2401-2500/2464_Minimum_Subarrays_in_a_Valid_Split/C/1_Approach_Dynamic_Programming.c
const int INF = 1e9;

int gcd(int a, int b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

int validSubarraySplit(int* nums, int numsSize) {
    int dp[numsSize + 1];
    for (int i = 1; i <= numsSize; i++) {
        dp[i] = INF;
    }
    dp[0] = 0;
    for (int i = 1; i <= numsSize; i++) {
        for (int j = 1; j <= i; j++) {
            if (gcd(nums[i - 1], nums[j - 1]) > 1) {
                dp[i] = (dp[i] < dp[j - 1] + 1) ? dp[i] : dp[j - 1] + 1;
            }
        }
    }
    return (dp[numsSize] == INF) ? -1 : dp[numsSize];
}
